File: 백준/2493.cpp
#include <bits/stdc++.h>

using namespace std;

int main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

    // <내 앞의 최대 높이, 인덱스>
    stack<pair<int, int>> S;

    // 첫번째 타워의 신호는 수신 불가
    S.push({100000001, 0});

    int n; cin >> n;
    for(int i=1; i<=n; i++){
        int num; cin >> num;

        // 입력받은 높이보다 큰 높이가 나올때까지 pop
        while(S.top().first < num) S.pop();

        // 큰 높이가 나오면 그때의 인덱스 출력하고 입력받은 높이, 인덱스 push
        cout << S.top().second << " ";
        S.push({num, i});
    }
}

File: 백준/2493.java
package org.elice;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Scanner;
import java.util.Stack;
import java.util.StringTokenizer;

class Pair{
    int topNumber;
    int topIndex;

    Pair(int topNumber, int topIndex){
        this.topNumber = topNumber;
        this.topIndex = topIndex;
    }
}

public class Main {
    public static void main(String[] args) throws IOException {
        Scanner sc = new Scanner(System.in);
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));


        int N = Integer.parseInt(br.readLine());
        StringTokenizer st = new StringTokenizer(br.readLine());

        Stack<Pair> stack = new Stack<>();
        stack.push(new Pair(100000001, 0));

        for(int i=1; i<=N; i++){
            int num = Integer.parseInt(st.nextToken());

            while(stack.peek().topNumber < num) stack.pop();

            System.out.print(stack.peek().topIndex + " ");
            stack.push(new Pair(num, i));
        }
    }
}